### ISSUE TITLE
exceptions and crashes..

### ISSUE THREAD

**Issue by macsmol** [dykstrom/ronja#14]
Hi. I've been testing my engine against ronja 0.8.2. Ronja lost 10/10 games due to crashes/exceptions. I thought you may want to know about it. Here are logs from Arena:
[ronja.0.8.2.arenaDebug.log](https://github.com/user-attachments/files/15789563/ronja.0.8.2.arenaDebug.log)

I didn't test the 0.9.0 version.

I played 10 game tournament with some predefined starting positions: 
```
[Event "1 .King's Pawn Game"]
1.e4 e5
[Event "2. Queen's Pawn Game"]
1.d4 
[Event "3. English Opening"]
1.c4 
[Event "4. Zukertort Opening"]
1.Nf3
[Event "5. Scandinavian Defense"]
1.e4 d5
```
I turned off ronjas book but then I tried to start it with book to - didn't make any difference.

**Comment by dykstrom** [comment]
Thanks for letting me know! Which version of Java were you using?

**Comment by tissatussa** [comment]
i got a notification of this Issue and i did a quick test with Ronja v0.9.0 - indeed the engine had some problems : see the PGNs below from CuteChess GUI. I always run one game at a time and leave CuteChess open, so all previous Ronja instances may have remained in memory .. is Ronja closed properly ?

btw. to run the engine i use your batch file which gives Ronja more memory then my default 128 Mb Hash, how can i change the parameters for that ? (i'm not into Java) Are the '-Xm' needed ? Does Ronja have a default for those ?

`java -Xms512M -Xmx1524M -jar ./ronja-0.9.0_xboard.jar`

```
$ java -version
openjdk version "21.0.3" 2024-04-16
OpenJDK Runtime Environment (build 21.0.3+9-Ubuntu-1ubuntu122.04.1)
OpenJDK 64-Bit Server VM (build 21.0.3+9-Ubuntu-1ubuntu122.04.1, mixed mode, sharing)
```

At first i let this Ronja version play a 7m3s game and all went well. Then i changed the TimeControl values, you'll see what happened :

(i stripped moves with "[..]" to unclutter the info)

[ i'm on Xubuntu 22.04 ]

```
[Event "engine vs engine"]
[Site "Holland"]
[Date "2024.06.11"]
[Round "?"]
[White "Ronja v0.9.0"]
[Black "ElephantGambit v0.7.0"]
[Result "0-1"]
[ECO "B03"]
[GameDuration "00:18:54"]
[Opening "Alekhine's defense"]
[PlyCount "136"]
[TimeControl "420+3"]

1. e4 Nf6 {-0.36/7 21s} 2. e5 Nd5 {-0.29/7 20s}
[..]
67. Kh2 {7.7s} Kf3 {/1 0s}
68. Kh3 {11s} Rh1# {/1 0s, Black mates} 0-1

[Event "engine vs engine"]
[Site "Holland"]
[Date "2024.06.11"]
[Round "?"]
[White "Ronja v0.9.0"]
[Black "4ku v4.0"]
[Result "0-1"]
[ECO "A02"]
[GameDuration "00:03:44"]
[Opening "Bird's Opening"]
[PlyCount "60"]
[TimeControl "60+3"]

1. f4 Nf6 {+0.40/20 2.8s} 2. Nf3 d5 {+0.34/20 3.1s}
[..]
30. Rg2 {2.9s} Qxg2# {+299.99/50 21s, Black mates} 0-1

[Event "engine vs engine"]
[Site "Holland"]
[Date "2024.06.11"]
[Round "?"]
[White "Ronja v0.9.0"]
[Black "ElephantGambit v0.7.0"]
[Result "0-1"]
[ECO "B02"]
[GameDuration "00:01:21"]
[Opening "Alekhine's defense"]
[PlyCount "4"]
[Termination "stalled connection"]
[TimeControl "60"]
[Variation "Mokele Mbembe (Buecker) variation"]

1. e4 Nf6 {-0.16/6 2.5s} 2. e5 Ne4 {-0.12/6 2.4s, White's connection stalls} 0-1

[Event "engine vs engine"]
[Site "Holland"]
[Date "2024.06.11"]
[Round "?"]
[White "Ronja v0.9.0"]
[Black "ElephantGambit v0.7.0"]
[Result "0-1"]
[ECO "B02"]
[GameDuration "00:03:30"]
[Opening "Alekhine's defense"]
[PlyCount "4"]
[Termination "stalled connection"]
[TimeControl "180"]
[Variation "Mokele Mbembe (Buecker) variation"]

1. e4 Nf6 {-0.36/7 7.5s} 2. e5 Ne4 {-0.12/6 7.2s, White's connection stalls} 0-1

[Event "engine vs engine"]
[Site "Holland"]
[Date "2024.06.11"]
[Round "?"]
[White "Ronja v0.9.0"]
[Black "ElephantGambit v0.7.0"]
[Result "0-1"]
[ECO "B03"]
[GameDuration "00:05:51"]
[Opening "Alekhine's defense"]
[PlyCount "6"]
[Termination "stalled connection"]
[TimeControl "300"]

1. e4 Nf6 {-0.36/7 13s} 2. e5 Nd5 {-0.29/7 12s} 3. d4
Nc6 {-0.11/7 11s, White's connection stalls} 0-1

[Event "engine vs engine"]
[Site "Holland"]
[Date "2024.06.11"]
[Round "?"]
[White "Ronja v0.9.0"]
[Black "ElephantGambit v0.7.0"]
[Result "0-1"]
[ECO "A02"]
[GameDuration "00:07:50"]
[Opening "Bird's Opening"]
[PlyCount "4"]
[Termination "stalled connection"]
[TimeControl "420"]

1. f4 e6 {-0.15/7 18s} 2. Nf3 Nf6 {-0.15/7 17s, White's connection stalls} 0-1
```

**Comment by tissatussa** [comment]
@macsmol : do you use a Mac ? Windows ? Linux ?
@dykstrom : did you only develop Ronja for Windows ?

**Comment by macsmol** [comment]
I dodn't develop Ronja. I'm developing my own engine - Magog. And I made it play against Ronja. And it crushed Ronja because Ronja had some bugs.
I don't think the bugs are related to java version.
If you look for the stack traces in the log file that I've attached you'll see that first problem was a division by 0:
```
2024-06-11 14:02:18,915-->2:time 18000
2024-06-11 14:02:18,915-->2:otim 17395
2024-06-11 14:02:18,915-->2:go
2024-06-11 14:02:18,930<--2:Exception in thread "main" java.lang.ArithmeticException: / by zero
2024-06-11 14:02:18,946<--2:	at se.dykstrom.ronja.engine.time.TimeUtils.calculateTimeForNextMove(TimeUtils.java:138)
2024-06-11 14:02:18,962<--2:	at se.dykstrom.ronja.engine.ui.command.AbstractMoveCommand.move(AbstractMoveCommand.java:62)
2024-06-11 14:02:18,983<--2:	at se.dykstrom.ronja.engine.ui.command.GoCommand.execute(GoCommand.java:45)
2024-06-11 14:02:18,999<--2:	at se.dykstrom.ronja.engine.ui.Ronja.start(Ronja.java:85)
2024-06-11 14:02:18,999<--2:	at se.dykstrom.ronja.engine.ui.Ronja.main(Ronja.java:63)
2024-06-11 14:05:19,966-->2:?
2024-06-11 14:05:19,982****Tour**Game end: 1-0 {1-0 Black forfeits on time}
2024-06-11 14:05:19,982****Tour**Last game won by Magog.0.9 with White in 2 moves
```
And another one was an attempt to parse  as timestamp something that apparently is not a time stamp:

```
2024-06-11 14:05:36,860-->2:time 17998
2024-06-11 14:05:36,860-->2:otim 16817
2024-06-11 14:05:36,860-->2:usermove d7d6
2024-06-11 14:05:36,884<--2:Exception in thread "main" java.time.DateTimeException: Invalid value for NanoOfDay (valid values 0 - 86399999999999): -1000000
2024-06-11 14:05:36,900<--2:	at java.base/java.time.temporal.ValueRange.checkValidValue(ValueRange.java:319)
2024-06-11 14:05:36,900<--2:	at java.base/java.time.temporal.ChronoField.checkValidValue(ChronoField.java:718)
2024-06-11 14:05:36,915<--2:	at java.base/java.time.LocalTime.ofNanoOfDay(LocalTime.java:400)
2024-06-11 14:05:36,931<--2:	at se.dykstrom.ronja.engine.time.TimeUtils.formatTime(TimeUtils.java:56)
2024-06-11 14:05:36,947<--2:	at se.dykstrom.ronja.engine.core.AlphaBetaFinder.findBestMoveWithinTime(AlphaBetaFinder.java:71)
2024-06-11 14:05:36,962<--2:	at se.dykstrom.ronja.engine.ui.command.AbstractMoveCommand.move(AbstractMoveCommand.java:64)
2024-06-11 14:05:36,982<--2:	at se.dykstrom.ronja.engine.ui.command.UserMoveCommand.execute(UserMoveCommand.java:67)
2024-06-11 14:05:36,982<--2:	at se.dykstrom.ronja.engine.ui.Ronja.start(Ronja.java:85)
2024-06-11 14:05:36,997<--2:	at se.dykstrom.ronja.engine.ui.Ronja.main(Ronja.java:63)
2024-06-11 14:08:37,950-->2:?
2024-06-11 14:08:37,957****Tour**Game end: 0-1 {0-1 White forfeits on time}
2024-06-11 14:08:37,957****Tour**Last game won by Magog.0.9 with Black in 4 moves
```

Hope this helps

**Comment by dykstrom** [comment]
I don't have access to a Windows computer right now, so I cannot reproduce the problem with Arena at the moment. But to answer the OS question, Ronja is supposed to work on Linux and macOS as well as Windows, since it is written in Java. When I can find some time I will try to reproduce the problems and see if I can fix them.

The -Xm parameters are not needed. They are parameters to the Java virtual machine, and Java has default values for them. Ronja itself does not use those parameters. And Ronja so far has no hash tables, so it does not care about has memory settings.

**Comment by dykstrom** [comment]
By the way, if someone has a Ronja log file (probably named ronja.log and written to Ronja installation directory), I would be happy if you could attach it to this issue.

**Comment by tissatussa** [comment]
> ..a Ronja log file (probably named ronja.log and written to Ronja installation directory)..

after running Ronja v0.9.0, i find no .log file in its folder ..

**Comment by dykstrom** [comment]
It seems that the root cause is that Ronja does not handle incremental time with a time increment of 0.

**Comment by tissatussa** [comment]
indeed v0.9.1 runs fine, also without time bonus.
and i have logging now, it seems i was missing the file 'ronja.properties', which is included in the v0.9.1 package.
thanks!

**Comment by dykstrom** [comment]
Great! Thanks for verifying!
